### ISSUE TITLE
Does not work with pg-pool

### ISSUE THREAD

**Issue by colideum** [voxpelli/node-connect-pg-simple#36]
Hi,
Module itself is great. Thank You. The only thing, it cannot work with pg-pool.
The only difference, i noticed, is in parameters of pg.connect. for pg-pool instance conString parameter is not needed (it is defined during initialization of pg-pool). As of myself i rewritten this function to use with my app, but basically it would be great, if there would be one of the options to use pg-pool,

**Comment by voxpelli** [comment]
Haven't looked at the `pg-pool` module, but if it's a simple feature to add support for then a PR would probably be accepted.

**Comment by Dill1** [comment]
@voxpelli, that would be very much appreciated. I'm also using pg-pool and would like to use this module.

**Comment by colideum** [comment]
as a workaround i did the following:
`	var sess = require('connect-pg-simple')(session);
	sess.prototype.query = function (query, params, fn) {
		if (!fn && typeof params === 'function') {
			fn = params;
		}
		this.pg.connect(function (err, client, done) {
			if (err) {
				done(client);
				if (fn) {
					fn(err);
				}
			} else {
				client.query(query, params || [], function (err, result) {
					done(err || false);
					if (fn) {
						fn(err, result && result.rows[0] ? result.rows[0] : false);
					}
				});
			}
		});
	}` Maybe it's not a correct way to do, but it works for me.

**Comment by sattellite** [comment]
+1

**Comment by illz** [comment]
Thanks Colideum - I refactored this slightly with pg-pool, ES6, and client.release() after the query executes
`
import session from 'express-session'
import pg from 'pg'
import Pool from 'pg-pool'
import pgSimpleSession from 'connect-pg-simple'

// This is configuration to use pg-pool connections for connect-pg-simple sessions
var poolCon = new Pool(Config.db)
// Override pgSession's query method with one that uses pg-pool connection
var pgSession = new pgSimpleSession(session)
pgSession.prototype.query = (query, params = [], callback) => {
  if(!callback && typeof params === 'function') {
    callback = params
    params = []
  }
  poolCon.connect().then(client => {
    client.query(query, params, (err, result) => {
      client.release()
      if(err) {
        return callback(err)
      } else {
        return callback(null, (result && result.rows[0] ? true : false))
      }
    })
  })
}
app.use(session({
  store: new pgSession({
    pg : poolCon, // pg-pool connection configured above
    tableName : 'sessions_studio_api'
  }),
  saveUninitialized: true,
  secret: Config.sessionSecret,
  resave: false,
  cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 } // 30 days
}));
`

**Comment by jWilko** [comment]
I've been looking for a way to get [connect-pg-simple](https://github.com/voxpelli/node-connect-pg-simple) work nicely with [pg-promise](https://github.com/vitaly-t/pg-promise).  Adapting the ideas above from @illz and @colideum (thanks both) I came up with this which may work for others:

server.js: 
```javascript
const session = require('express-session');
...
app.use(session({
        store: services.database.getPostgresSessionStore(session),
        secret: '****',
        saveUninitialized: false,
        resave: false,
        rolling: true,
        cookie: {
            ...
        }
    }));
```

services/Database.js:
```javascript
const pgp = require('pg-promise')();
const ConnectPgSimple = require('connect-pg-simple');

class Database {
    constructor(dbConfig) {
        this.db = pgp(dbConfig);
    }

    query(queryList) {
        ...
    }

    getPostgresSessionStore(session) {
        const pgSession = new ConnectPgSimple(session);
        pgSession.prototype.query = (query, params = [], fn) => {
            if (!fn && typeof params === 'function') {
                fn = params;
                params = [];
            }
            this.db.query(query, params)
                .then( result => {
                    if (fn) {
                        fn(null, result && result[0]);
                    }
                })
                .catch( err => {
                    if (fn) {
                        fn(err);
                    }
                });
        };
        return new pgSession({
            pg: this.db
        });
    }

}

module.exports = Database;
```

There may be a simpler solution out there.  Please shout if you have one.

**Comment by eugene1g** [comment]
I did a slightly different take to mocking the `pg` public interface like so - 
```
const _mockQueryFn = (queryString, params, queryCallback) => {
  pgPromiseInstance //<--- here put your reference to the pg-promise instance for your database
    .query(queryString, params)
    .then(result => queryCallback(null, { rows: result }))
    .catch(error => queryCallback(error));
};
const _mockPgObject = {
  connect: (conString, connectCallback) => {
    connectCallback(null, { query: _mockQueryFn }, () => {});
  }
};
// .... config in the session init

app.use(
    session({
      store: new pgSession({
        pg: _mockPgObject
      }),
      secret: "....",
      resave: false,
      saveUninitialized: false,
      cookie: { maxAge: 30 * 24 * 60 * 60 * 1000 }
    })
  );

```

**Comment by voxpelli** [comment]
This was very long overdue, sorry for that, but now it's fixed and released as `4.0.0` 🎉 Many thanks to @i8-pi for helping out!
